larpandoracontent/LArReclustering/CheatedThreeDClusteringTool.cc

```diff
@@ -0,0 +1,109 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/CheatedThreeDClusteringTool.cc
+ *
+ *  @brief  Implementation file for the reclustering algorithm that uses transverse calorimetric profiles.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+
+#include "larpandoracontent/LArReclustering/CheatedThreeDClusteringTool.h"
+
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+CheatedThreeDClusteringTool::CheatedThreeDClusteringTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool CheatedThreeDClusteringTool::Run(const Algorithm *const pAlgorithm, std::vector<pandora::CaloHitList*> &newCaloHitListsVector)
+{
+    if (newCaloHitListsVector.empty() || newCaloHitListsVector.size() != 1)
+        return false;
+
+    CaloHitList *initialCaloHitList = newCaloHitListsVector.at(0);
+    newCaloHitListsVector.clear();
+    for (const CaloHit *const pCaloHit : *initialCaloHitList)
+    {
+        const CaloHit *const pParentCaloHit = static_cast<const CaloHit *>(pCaloHit->GetParentAddress());
+
+        int mainMcParticleIndex = this->GetMainMcParticleIndex(pAlgorithm, pParentCaloHit);
+
+        // Flag to indicate if a matching list was found
+        bool found = false;
+
+        // Search for an existing list with matching MC particle index
+        for (CaloHitList *caloHitList : newCaloHitListsVector)
+        {
+            const CaloHit *const pListParentCaloHit = static_cast<const CaloHit *>(caloHitList->front()->GetParentAddress());
+            int listMainMcParticleIndex = this->GetMainMcParticleIndex(pAlgorithm, pListParentCaloHit);
+            if (listMainMcParticleIndex == mainMcParticleIndex)
+            {
+                found = true;
+                caloHitList->push_back(pCaloHit);
+                break;
+            }
+        }
```

Comment:
This is perhaps just a different way of approaching the same problem, but I feel like re-using the `newCaloHitListsVector` here is not the optimal way to do it.

If I'm understanding correctly, the code is broadly: For every calo hit, get its MC, check if that MC ID is already stored, and if it is add the hit to the existing list, if not, add a new list.

It feels like a slightly easier way of representing that would just be with a `std::map<int, CaloHitList>`, so you can skip the inner loop and just do a direct look up of `if mcMap.contains(mcIndex)` and then go from there. Would get rid of the loop, the flag variable etc.

Reply:
Thanks Ryan! I am pushing a new version that uses a map instead of the inner loop. However, I still reuse newCaloHitListsVector, which initially contains the list of merged calo hits, and at the end contains all the new lists that have been created using cheated information. This then gets passed to and used by the main reclustering algorithm.